### DISCUSSION TITLE
IMPORTANT: Fixing flaws in SAF spec v1.0.9 and v2.0.0

### DISCUSSION THREAD

**Discussion by scia-nv** [StructuralAnalysisFormat/StructuralAnalysisFormat-Doc#7] [General]
Dear SAF implementors,

Over the past weeks, some issues have been raised regarding some of the latest changes (version 1.0.9 and 2.0.0) on MS teams.
There haven’t been many responses, so I'm reaching out to you, hoping for more feedback before making any changes.

I can't stress this enough, that these are changes to be made to the already published SAF specification for 1.0.9 and 2.0.0.
This means that anyone who might have already implemented these versions, might need to make some changes.

### • **Issue 1: Multi-type reference columns have been introduced**
These are columns in which references are made to multiple other types, by name.
In the past we've always had dedicated reference columns per type, so this is inconsistent with the rest of the specification.

Affected objects are:
o	**_StructuralCurveAction - "2D member" column._**
The "Force action" enumeration is updated in 2.0.0 with the values "On subregion edge" and "On opening edge", indicating that the "2D member" column will either contain the name of a _StructuralSurfaceMember,_ _StructuralSurfaceMemberRegion_ or _StructuralSurfaceMemberOpening_

**Proposed change:**
Add 2 new columns, "2D member region" and "2D member opening" which will act as reference columns to _StructuralSurfaceMemberRegion_ and _StructuralSurfaceMemberOpening_ respectively

o	**_StructuralCurveMoment - "2D member" column._**
The "Force action" enumeration is updated in 2.0.0 with the values "On subregion edge" and "On opening edge", indicating that the "2D member" column will either contain the name of a _StructuralSurfaceMember_, _StructuralSurfaceMemberRegion_ or _StructuralSurfaceMemberOpening_

**Proposed change:**
Add 2 new columns, "2D member region" and "2D member opening" which will act as reference columns to _StructuralSurfaceMemberRegion_ and _StructuralSurfaceMemberOpening respectively_

o	**_StructuralEdgeConnection - "2D member" column._**
In 2.0.0, a new "Boundary condition" column has been added, which would allow the "2D member" column to act as a reference column to either _StructuralSurfaceMember_, _StructuralSurfaceMemberRegion_ or _StructuralSurfaceMemberOpening_.

**Proposed change:** 
Add 2 new columns, "2D member region" and "2D member opening" which will act as reference columns to _StructuralSurfaceMemberRegion_ and _StructuralSurfaceMemberOpening_ respectively
                  
### •	Issue 2: _StructuralSurfaceActionDistribution_ name too long use as worksheet name
In 1.0.9 a new object called _StructuralSurfaceActionDistribution_ was added, but the name _StructuralSurfaceActionDistribution_ is too long to be used as worksheet name.
Excel has a limit of 31 characters for worksheet names, so we cannot use _StructuralSurfaceActionDistribution_.

**Proposed change:** 
Set the worksheet name to "**StructuralSurfaceActionDistri"**

### •	Issue 3: Fundamental issue with reference to _StructuralSurfaceActionDistribution_ on _StructuralSurfaceAction_
In 1.0.9, with the addition of _StructuralSurfaceActionDistribution_, an option was added to the "2D member" column on _StructuralSurfaceAction_ object to also reference _StructuralSurfaceActionDistribution_. 
While this issue is similar to **Issue 1: Multi-type reference columns have been introduced**, it has an additional flaw being that it is impossible to determine if the name in the "2D member" column refers to the name of a _StructuralSurfaceMember_ or _StructuralSurfaceActionDistribution_ object.
Imagine having both of these objects in your model with the same name!

**Proposed change:**
Add new "2D member distribution" column which acts as the reference to _StructuralSurfaceActionDistribution_

**Optional additional change:** 
Add a new enumeration, similar to the "Force action" enumeration which will specify which kind of type the _StructuralSurfaceAction_ references

Could you kindly please reply whether or not you agree with the proposed changes before the 23rd of July?
If I haven’t heard back from you by then, I’ll assume you agree with the proposed changes. Discussion is open here.

**Comment by reischlb** [comment]
Hello,

Just a note to the third issue: I assumed that names are globally unique so you can't have two object with the same name neither with different types, since the IDs are not mandatory.
I am not sure where did I read about it, I think it was on the previous saf.guide or may it was just an assumption from me.
So what about this?

Globally unique names will solve both issues 1 and 3, because than StructuralCurveAction, StructuralCurveMoment etc. don't have to know what is exactly the underlying object, it's enough that there are something which can act as a structural object's edge.

I think this would make implementations easier and it would eliminate that erroneous case when you set the value of the "force action" enum but enter the underlying object's name in a wrong cell.

**Reply by dschuermans** [reply]
Hey, we have always worked with names being unique **within their own type**, but this was never explicitly specified in the documentation. 
All of the object pages have as description for the Name column **"Human readable unique name of the ..."**
This does leave it open for how you want to interpret it, be it unique within their type or unique within the model.
The IDs are however, unique within the entire model.

But if you've always worked under the assumption that names are unique across the entire model, even better!

The force action kind of columns might look like a bit overkill indeed, but it does open up the ability to do more specific validations on your model and provide the user with better feedback. _(which I believe was also the goal at the time, when we first introduced these sort of columns)_

With the current setup _(force action column + dedicated column for each type of reference)_, you are able to generate messages like "You need to provide the name of ... because Force Action is set to ..."
It also allows you to ignore other references automatically, eg. if 2 reference columns have a value, you can ignore the value from the column which doesn't match with the force action value.
If the user then does enter a name of an object of a different type, you'll also be able to generate a message such as "No object of type ... was found with the name ..."

However, if you do not have a force action column, you'll have to use more generic messages such as:
* You need to provide the name of ... or ... or ...
* You can only provide 1 name for the reference, when the user has entered multiple
* The name ... was not found as an ... or ... or ...
You are also limiting your ability to do some automatic processing. 
You won't be able to ignore a column when more than 1 name was provided, because you won't know which one the user intended to use.

Then from a technical point-of-view, which would be more performant you think?
* Having a list with all objects that could be referenced, iterating over each object to see if the name matches?
* Having a pre-filtered list of objects of a single type and iterating over them to see if the name matches?

But yes, globally unique names would have solved the issue if you only take into account the issue itself and not everything around it. 
Keep in mind there's an SDK now that's been built along with the spec from the start, in which we can't simply throw everything overboard and refactor a lot of things. We simply don't have the time or resources for that.

**Reply by Balinth** [reply]
Hello,

full disclosure: I am from the same team as @reischlb .

We of course know the constricting effects of time limitations on development. I also see that these proposed changes have gone through but would still like to answer to your points, maybe as something thought provoking for SAF 3.0:

One of the better features of any kind of Interop format such as SAF is to make as much of the incorrect object states unrepresentable as possible with the data representation structure.

Eg.: you say that without the force action column we lose the ability to automatically process which reference columns should we ignore.
But really, we lose the **need** to process this, because ideally there would be only one column, that accepts the possible types through a polymorphic interface type. This way, the user / engineer / exporter can never make the mistake of exporting with 2 different references, one of which will be ignored later during import. 
(note also, the engineers need not know what "polymorphic interface type" mean, they only need the basic understanding that this column can reference any kind of structural object that has a 2d outline, be it a simple 2d member, a region on a member, or a hole, we use OOP because it is intuitive after all)

As for error reporting, yes, we would think it is much more useful to report to the user "This column must reference one of these kinds of objects: 2d member, region, opening". This is direct and actionable information.
In contrast "You need to provide the name of ... because Force Action is set to ..." while not unhelpful, still requires the user to jump though additional hoops, looking up "what are the possible values of Force Action?".
Not to mention, here we really loose the ability to do some automatic processing for them. Why should the exporter need to specify the Force Action enum manually, when we could perfectly easily figure it out during the import process, provided the names are globally unique?

In any kind of error reporting relating to a specific object, which is better, if there are multiple objects with the same name leaving the user to take additional context into account in finding the problematic object, or if there is always just one object, that is easily identifiable through its name?

Finally from the technical point of view, especially regarding performance, why on earth would we iterate over lists where the size is significant for lookup? Obviously we just use a hashmap of the names and objects, with approx constant lookup time, without the need to do prefilterings etc.

Sorry for the ranty thoughts, but in my experience, globally unique names would be exactly the better systematic solution considering the wider implications, which is why it never even occurred to us that the names are only unique per type.

**Reply by dschuermans** [reply]
No problem at all, that's what discussions are for 👍 

I've initially tried to introduce these changes as they were defined into the SDK, but it simply was not do-able at this time.
Those changes were introduced to the specification without checking the feasibility with an actual software developer that has to implement it in the SDK ☹️ 
It would've required too much effort to implement

But at least now we'll have a trail of some new ideas for the next version 😉

**Comment by sonkaste17** [comment]
Hi everyone,
according to collected feedback with no disagreement, we will do the proposed changes in SAF 2.0.0 and SAF 1.0.9

- ### **Issue 1** : Multi-type reference columns have been introduced.

    - **_StructuralCurveAction_** : Add 2 new columns, "2D member region" and "2D member opening" which will act as reference columns to StructuralSurfaceMemberRegion and StructuralSurfaceMemberOpening respectively (SAF 2.0.0)
    - **_StructuralCurveMoment_** : Add 2 new columns, "2D member region" and "2D member opening" which will act as reference columns to StructuralSurfaceMemberRegion and StructuralSurfaceMemberOpening respectively (SAF 2.0.0)
    - **_StructuralEdgeConnection_** : Add 2 new columns, "2D member region" and "2D member opening" which will act as reference columns to StructuralSurfaceMemberRegion and StructuralSurfaceMemberOpening respectively (SAF 2.0.0)

- ### Issue 2:  StructuralSurfaceActionDistribution name too long use as worksheet name
Set the worksheet name to "StructuralSurfaceActionDistri" in SAF 1.0.9

- ### Issue 3: Fundamental issue with reference to StructuralSurfaceActionDistribution on StructuralSurfaceAction
Add new "2D member distribution" column which acts as the reference to StructuralSurfaceActionDistribution in SAF 1.0.9

Add a new enumeration, similar to the "Force action" enumeration which will specify which kind of type the StructuralSurfaceAction references
